src/chainstate/stacks/boot/cost-voting.clar

```diff
@@ -0,0 +1,205 @@
+(define-read-only (get-principal-votes (address principal) (proposal-id uint))
+    (get votes (map-get? principal-proposal-votes { address: address, proposal-id: proposal-id })))
+
+;; Propose cost-functions
+(define-public (submit-proposal (function-contract principal)
+                                (function-name (string-ascii 128))
+                                (cost-function-contract principal)
+                                (cost-function-name (string-ascii 128)))
+    (begin
+        (map-insert proposals { proposal-id: (var-get proposal-count) }
+                              { cost-function-contract: cost-function-contract,
+                                cost-function-name: cost-function-name,
+                                function-contract: function-contract,
+                                function-name: function-name,
+                                expiration-block-height: (+ burn-block-height u2016) })
+        (map-insert proposal-votes { proposal-id: (var-get proposal-count) } { votes: u0 })
+        (var-set proposal-count (+ (var-get proposal-count) u1))
+        (ok (- (var-get proposal-count) u1))))
+
+;; Vote on a proposal
+(define-public (vote-proposal (proposal-id uint) (amount uint))
+    (let (
+        (cur-votes (default-to u0 (get votes (map-get? proposal-votes { proposal-id: proposal-id }))))
+        (cur-principal-votes (default-to u0 (get votes (map-get? principal-proposal-votes {
+            address: tx-sender,
+            proposal-id: proposal-id }))))
+        (expiration-block-height (get expiration-block-height (unwrap! (map-get? proposals {
+            proposal-id: proposal-id }) (err ERR_NO_SUCH_PROPOSAL))))
+    )
+
+    (asserts! (> amount u0) (err ERR_AMOUNT_NOT_POSITIVE))
+    (asserts! (< burn-block-height expiration-block-height) (err ERR_PROPOSAL_EXPIRED))
+    (asserts! (is-none (map-get? proposal-confirmed-id { proposal-id: proposal-id }))
+        (err ERR_PROPOSAL_CONFIRMED))
+
+    (unwrap! (stx-transfer? amount tx-sender (as-contract tx-sender)) (err ERR_INSUFFICIENT_FUNDS))
+    (unwrap! (ft-mint? cost-vote-token amount tx-sender) (err ERR_FT_MINT))
+
+    (map-set proposal-votes { proposal-id: proposal-id } { votes: (+ amount cur-votes) })
+    (map-set principal-proposal-votes { address: tx-sender, proposal-id: proposal-id}
+                                    { votes: (+ amount cur-principal-votes)})
+    (ok true))
+)
+
+;; Withdraw votes
+(define-public (withdraw-votes (proposal-id uint) (amount uint))
+    (let (
+        (cur-votes (default-to u0 (get votes (map-get? proposal-votes { proposal-id: proposal-id }))))
+        (cur-principal-votes (default-to u0 (get votes (map-get? principal-proposal-votes {
+            address: tx-sender,
+            proposal-id: proposal-id }))))
+        (expiration-block-height (get expiration-block-height (unwrap! (map-get? proposals {
+            proposal-id: proposal-id }) (err ERR_NO_SUCH_PROPOSAL))))
+        (sender tx-sender)
```

Comment:
Slight optimization -- you can remove `(sender tx-sender)` and just use `tx-sender`.